src/JSInterop/Microsoft.JSInterop/src/DotNetObjectRefOfT.cs

```diff
@@ -0,0 +1,117 @@
+        public DotNetObjectRef(TValue value)
+        {
+            Value = value;
+            _trackingId = DotNetObjectRefManager.Current.TrackObject(this);
+        }
+
+        /// <summary>
+        /// Gets the object instance represented by this wrapper.
+        /// </summary>
+        [JsonIgnore]
+        public TValue Value
+        {
+            get;
+            // Workaround for https://github.com/dotnet/corefx/issues/37536 and https://github.com/dotnet/corefx/issues/37567
+            // Once fixed, we should make this property private.
+            set;
+        }
+
+        internal bool Disposed { get; private set; }
+
+        /// <summary>
+        /// This constructor is for meant for JSON serialization and should not be used by user code.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [JsonPropertyName(DotNetDispatcher.DotNetObjectRefKey)]
+        [Obsolete("This API is meant for JSInterop infrastructure and should not be used by user code.")]
+        public long TrackingId
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return _trackingId.Value;
+            }
+            set
+            {
+                ThrowIfDisposed();
+                if (_trackingId != null)
+                {
+                    throw new InvalidOperationException($"{nameof(DotNetObjectRef<TValue>)} cannot be reinitialized.");
+                }
+
+                _trackingId = value;
+                Value = (TValue)DotNetObjectRefManager.Current.FindDotNetObject(value);
+            }
+        }
+
+        object IDotNetObjectRef.Value => Value;
+
+        /// <summary>
+        /// Implictly converts the <see cref="DotNetObjectRef{TValue}" /> instance to the wrapping value.
+        /// </summary>
+        /// <param name="result">The <see cref="DotNetObjectRef{TValue}"/>.</param>
+        public static implicit operator TValue(DotNetObjectRef<TValue> result) => result.Value;
+
+        /// <summary>
+        /// Stops tracking this object reference, allowing it to be garbage collected
+        /// (if there are no other references to it). Once the instance is disposed, it
+        /// can no longer be used in interop calls from JavaScript code.
+        /// </summary>
+        public void Dispose()
+        {
+            if (!Disposed)
+            {
+                Disposed = true;
+                // Make sure the object was constructed correctly.
+                DotNetObjectRefManager.Current.ReleaseDotNetObject(_trackingId.Value);
+            }
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (Disposed)
+            {
+                throw new ObjectDisposedException(nameof(DotNetObjectRef<TValue>));
+            }
+        }
+
+        // Used by the RefManager to mark this instance as disposed once the object is no longer being tracked.
+        void IDotNetObjectRef.SetDisposed()
```

Comment:
Could you clarify why we have this feature? I'm not sure what bad thing would happen if we let the `DotNetObjectRefManager` deal with disposal, and the `DotNetObjectRef<T>` wouldn't need to know whether it had been disposed.

Is it to stop people from passing the value to JS after it's already been disposed? Because I'm not sure what that protects people from, as their JS code can already hold onto instances after disposal and try to use them regardless. Is there another scenario I'm missing?

Reply:
> Is it to stop people from passing the value to JS after it's already been disposed

Sort of. But yeah, it seems unnecessary. Cleaning it up doe